src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedEntryPointSymbol.cs

```diff
@@ -329,6 +334,160 @@ private static BoundCall CreateParameterlessCall(CSharpSyntaxNode syntax, BoundE
+
+            private readonly MethodSymbol _getAwaiterMethod;
+            private readonly MethodSymbol _getResultMethod;
+
+            private readonly ImmutableArray<ParameterSymbol> _parameters;
+
+            // Which "kind" of Main is this: `void Main` or `int Main`?
+            // This is determined by looking at the return type of the user-provided Main
+            private enum ReturnKind
+            {
+                VoidReturning,
+                IntReturning,
+            }
+            // Does this main method need to forward arguments to the user-provided Main?
+            private enum ParamKind
+            {
+                NoArgs,
+                StringArrayArgs,
+            }
+
+            // Task -> Void
+            // Task<_> -> Int32
+            private static TypeSymbol TranslateReturnType(CSharpCompilation compilation, TypeSymbol returnType)
+            {
+                return returnType.IsGenericTaskType(compilation)
+                    ? compilation.GetSpecialType(SpecialType.System_Int32)
+                    : compilation.GetSpecialType(SpecialType.System_Void);
+            }
+
+            // Task -> ReturnKind.VoidReturning
+            // Task<_> -> ReturnKind.IntReturning
+            private static ReturnKind GetReturnKind(CSharpCompilation compilation, TypeSymbol returnType)
+            {
+                return returnType.IsGenericTaskType(compilation) ? ReturnKind.IntReturning : ReturnKind.VoidReturning;
+            }
+
+            internal AsyncForwardEntryPoint(CSharpCompilation compilation, DiagnosticBag diagnosticBag, NamedTypeSymbol containingType, MethodSymbol userMain) :
+                base(containingType, TranslateReturnType(compilation, userMain.ReturnType))
+            {
+                // There should be no way for a userMain to be passed in unless it already passed the 
+                // parameter checks for determining entrypoint validity.
+                Debug.Assert(userMain.ParameterCount == 0 || userMain.ParameterCount == 1);
+
+                _userMain = userMain;
+                _paramKind = userMain.ParameterCount == 0 ? ParamKind.NoArgs : ParamKind.StringArrayArgs;
+                _returnKind = GetReturnKind(compilation, userMain.ReturnType);
+
+                switch (_paramKind)
+                {
+                    case ParamKind.StringArrayArgs:
+                        var stringType = compilation.GetSpecialType(SpecialType.System_String);
+                        var stringArrayType = ArrayTypeSymbol.CreateCSharpArray(compilation.Assembly, stringType);
+                        _parameters = ImmutableArray.Create(
+                            SynthesizedParameterSymbol.Create(this, stringArrayType, 0, RefKind.None, userMain.Parameters[0].Name));
+                        break;
+                    default:
+                        _parameters = ImmutableArray<ParameterSymbol>.Empty;
+                        break;
```

Comment:
Consider separate cases for `case 0:` and `default:` and `throw UnexpectedValue()` in the latter. That would replace the `Debug.Assert(...ParameterCount)` above.